IA shuffle and edit of Contribute section (#13998)
PrefectHQ/prefect: docs/3.0rc/community/styles-practices.mdx

---
title: Styles and practices

> description: Prefect's 

Reviewer: ```suggestion
```

### Imports in modules

#### Importing other modules

The `from` syntax is recommended for importing objects from modules. You should not import modules 
with the `from` syntax.

```python
# Correct
import prefect.server.schemas  # use with the full name
import prefect.server.schemas as schemas  # use the shorter name

```

```python
# Wrong
from prefect.server import schemas

```

You should not use relative imports unless it's in an `__init__.py` file.

```python
# Correct
from prefect.utilities.foo import bar

```

```python
# Wrong
from .utilities.foo import bar

```

You should never use imports that are dependent on file location without explicit indication it is relative. This avoids 
confusion about the source of a module.

```python
# Correct
from . import test

```

#### Resolving circular dependencies

Sometimes, you must defer an import and perform it _within_ a function to avoid a circular dependency:

```python
## This function in `settings.py` requires a method from the global `context` but the context
## uses settings
def from_context():
    from prefect.context import get_profile_context

    ...

```

> Attempt to avoid circular dependencies. This often reveals entanglement in the design.

Reviewer: ```suggestion
Avoid circular dependencies. They often reveal entanglement in the design.
```

> You should place all deferred imports at the top of the function.

Reviewer: ```suggestion
Place all deferred imports at the top of the function.
```

    prefect work-queue inspect 'abcde'

```

Additionally:

*   Wrap generated arguments in apostrophes (') to ensure validity by using suffixing formats with `!r`.
*   Indent example commands, instead of wrapping in backticks (\`).
*   Use placeholders if you cannot completely pre-format the example.
*   Capitalize placeholder labels and wrap them in less than (\<) and greater than (>) signs.
*   Utilize `textwrap.dedent` to remove extraneous spacing for strings with triple quotes (""").

Placeholder Example:

```
Create a work queue with tags:
    prefect work-queue create '<WORK QUEUE NAME>' -t '<OPTIONAL TAG 1>' -t '<OPTIONAL TAG 2>'

```

Dedent Example:

```python
from textwrap import dedent
...
output_msg = dedent(
    f"""
    Created work queue with properties:
        name - {name!r}
        uuid - {result}
        tags - {tags or None}
        deployment_ids - {deployment_ids or None}

    Start an agent to pick up flows from the created work queue:
        prefect agent start -q {name!r}

    Inspect the created work queue:
        prefect work-queue inspect {name!r}
    """
)

```

## API versioning

### Client and server communication

> The Prefect client can be run separately from the Prefect orchestration server and communicate entirely via an API. 

Reviewer: ```suggestion
The Prefect client can be run separately from Prefect server and communicate entirely via an API. 
```

Among other things, the Prefect client includes anything that runs task or flow code, (e.g. agents, and the Python client) 

> or any consumer of Prefect metadata, (e.g. the Prefect UI, and CLI). The Prefect server stores this metadata and serves it

Reviewer: ```suggestion
or any consumer of Prefect metadata, (e.g. the Prefect UI, and CLI). Prefect server stores this metadata and serves it
```

via the REST API.

### API version header

> Sometimes we have to make breaking changes to the API (for good reasons). To check a Prefect client's compatibility

Reviewer: ```suggestion
Sometimes, we have to make breaking changes to the API. To check a Prefect client's compatibility
```

with the API it's making requests to, every API call the client makes includes a three-component `API_VERSION` header with major, 
minor, and patch versions.

For example, a request with the `X-PREFECT-API-VERSION=3.2.1` header has a major version of `3`, minor version `2`, and patch 
version `1`.

Change this version header by modifying the `API_VERSION` constant in `prefect.server.api.server`.

### Breaking changes to the API

> A breaking change means that your code needs to change to use a new version of Prefect. We strive to avoid breaking 

Reviewer: ```suggestion
A breaking change means that your code needs to change to use a new version of Prefect. We avoid breaking 
```

> changes in all releases.

Reviewer: ```suggestion
changes whenever possible.
```

This means that the previous version of the client can still make calls against the updated version of the server code. 
This might happen if the changes are purely additive, such as adding a non-critical API route. In these cases, we aim 
to bump the patch version.

In almost all other cases, we bump the minor version, which denotes a non-backwards-compatible API change. We have reserved 
the major version changes to denote a backwards compatible change that is significant in some way, such as a major release 
milestone.

### Version composition

Versions are composed of three parts: MAJOR.MINOR.PATCH. For example, the version 2.5.0 has a major version of 2, a minor version 
of 5, and patch version of 0.

Occasionally, we add a suffix to the version such as `rc`, `a`, or `b`. These indicate pre-release versions that users can 
opt into to test functionality before it is ready for release.

Each release will increase one of the version numbers. If we increase a number other than the patch version, the versions to the 
right of it reset to zero.

## Prefect's versioning scheme

> Prefect increases the major version when significant and widespread changes are made to the core product. It is very 

Reviewer: ```suggestion
Prefect increases the major version when significant and widespread changes are made to the core product.
```

> unlikely that the major version will change without extensive warning.

Reviewer: ```suggestion
```
